lib/THC/THCDeviceTensor.cuh

```diff
@@ -0,0 +1,502 @@
+#ifndef THC_DEVICE_TENSOR_INC
+#define THC_DEVICE_TENSOR_INC
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+// A CUDA 6.5 compatible version of static_assert. Remove once on CUDA 7.0.
+template <bool>
+struct THCStaticAssert;
+
+template <>
+struct THCStaticAssert<true> {
+};
+
+#define thc_static_assert(expr) (THCStaticAssert<(expr) != 0>())
+
+/// Our tensor type
+template <typename T,
+          int Dim,
+          typename IndexT,
+          template <typename U> class PtrTraits>
+class THCDeviceTensor;
+
+/// Type of a subspace of a tensor
+namespace detail {
+template <typename TensorType,
+          int SubDim,
+          template <typename U> class PtrTraits>
```

Comment:
This was written against C++03, there is no TensorType::PtrTraits because you can't typedef a template. In C++11 you could say something like `template <typename U> using PtrTraits = PtrTraits<U>;` in THCDeviceTensor and use it that way.

Another C++03 option would have been to add an extra level of indirection, like

```
class THCDeviceTensor {
  ...
  template <typename U> struct UsingTraits { typedef PtrTraits<U> type; };
```

I just chose to template the sub-class on the traits in order to use it.